### PR TITLE
Agent scheduling

### PR DESCRIPTION
**Pull request by anisha1607**
### Description
Agent Scheduling - With this feature, the user can schedule agents for a particular day and time or schedule recurring agents for daily, weekly and monthly.

### Solution and Design
A new table of Agent Scheduler has been made which stores data of each schedule of the agent.
Endpoints have been made to create and schedule an agent, schedule an existing agent, edit a schedule, stop a schedule and get data of a schedule of an agent.
When a schedule is created for an agent, its status is set to SCHEDULED, when it is stopped, its status is set to STOPPED.When an agent completes its duration successfully, its status is set to COMPLETED, else TERMINATED.
Celery Beats has been used to schedule periodic tasks at an interval of 5 minutes. Each run checks if any agent which is running has a scheduled time in the past 5 minutes and takes those tasks and run it.

### Test Plan
For testing, pytest has been used which is one of the most common tools used for unit tests in Python. It is simple and flexible, and it offers easy no-boilerplate testing as well as powerful features for advanced users. 
Unit tests have been structured as functions that start with `test_` for all the functions that I have written. They will be found in the test folder.

### Type of change
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Docs update

### Checklist
- [x] My pull request is atomic and focuses on a single change.
- [x] I have read the contributing guide and my code conforms to the guidelines.
- [x] I have documented my changes clearly and comprehensively.
- [x] I have added the required tests.

### CODE REVIEW

**Review by luciferlinx101 — DISMISSED** [review]
Make changes

File: migrations/versions/ba86222087d1_agent_scheduler_table_created.py [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+
+
+# revision identifiers, used by Alembic.
+revision = 'ba86222087d1'
+down_revision = '7a3e336c0fba'
+branch_labels = None
+depends_on = None
+
+
+def upgrade() -> None:
```
Comment (TransformerOptimus) [inline migrations/versions/ba86222087d1_agent_scheduler_table_created.py:19]: rename it to create_agent_scheduler.py

File: migrations/versions/ba86222087d1_agent_scheduler_table_created.py [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+    sa.Column('recurrence_interval', sa.String(), nullable=True),
+    sa.Column('expiry_date', sa.DateTime(), nullable=True),
+    sa.Column('expiry_runs', sa.Integer(), nullable=True),
+    sa.Column('current_runs', sa.Integer(), nullable=True),
+    sa.Column('status', sa.String(), nullable=True),
+    sa.PrimaryKeyConstraint('id')
+    )
+    op.drop_index('ix_agent_execution_permissions_agent_execution_id', table_name='agent_execution_permissions')
+    op.drop_index('ix_atc_agnt_template_id_key', table_name='agent_template_configs')
+    op.drop_index('ix_agt_agnt_name', table_name='agent_templates')
```
Comment (TransformerOptimus) [inline migrations/versions/ba86222087d1_agent_scheduler_table_created.py:37]: why drop index?

Add more index on expiry_date, status, agent_id

File: superagi/controllers/agent.py [hunk 3]
```diff
@@ -119,6 +123,7 @@ def update_agent(agent_id: int, agent: sqlalchemy_to_pydantic(Agent, exclude=["i
 
 @router.post("/create", status_code=201)
 def create_agent_with_config(agent_with_config: AgentWithConfig,
+                             should_execute: bool = True,
```
Comment (TransformerOptimus) [inline superagi/controllers/agent.py:126]: it should be is_scheduled.

File: superagi/controllers/agent.py [hunk 4]
```diff
@@ -162,22 +168,217 @@ def create_agent_with_config(agent_with_config: AgentWithConfig,
...
+        model=agent_with_config_and_schedule.model,
+        permission_type=agent_with_config_and_schedule.permission_type,
+        LTM_DB=agent_with_config_and_schedule.LTM_DB,
+        memory_window=agent_with_config_and_schedule.memory_window,
+        max_iterations=agent_with_config_and_schedule.max_iterations,
+        user_timezone= agent_with_config_and_schedule.user_timezone
+    )
+
+    # First, create the agent with config but do not execute it
+    agent_data = create_agent_with_config(agent, should_execute=False)
```
Comment (TransformerOptimus) [inline superagi/controllers/agent.py:255]: controller cant call controller. Call model method or segregate the method into separate method.

File: superagi/models/agent_scheduler.py [hunk 5]
```diff
@@ -0,0 +1,66 @@
+from sqlalchemy import Column, Integer, String, Date, DateTime
+from superagi.models.base_model import DBBaseModel
+from superagi.models.types.agent_schedule import AgentScheduleCreate
+
+class AgentScheduler(DBBaseModel):
```
Comment (TransformerOptimus) [inline superagi/models/agent_scheduler.py:5]: change the table name to AgentSchedule.

File: superagi/models/tool.py [hunk 6]
```diff
@@ -100,3 +101,11 @@ def convert_tool_ids_to_names(cls, db, tool_ids):
 
         tools = db.session.query(Tool).filter(Tool.id.in_(tool_ids)).all()
         return [str(tool.name) for tool in tools]
+    
+    @staticmethod
+    def is_tool_id_valid(agent_with_config, session):
```
Comment (TransformerOptimus) [inline superagi/models/tool.py:106]: this method does not look complete.

File: superagi/controllers/agent.py [hunk 7]
```diff
@@ -178,6 +173,123 @@ def create_agent_with_config(agent_with_config: AgentWithConfig,
...
+    agent_toolkit_tools = AgentConfiguration.get_tools_from_agent_config(session=db.session,
+                                                                         agent_with_config=agent_with_config_and_schedule.agent)
+    agent_with_config_and_schedule.agent.tools.extend(agent_toolkit_tools)
+    db_agent = Agent.create_agent_with_config(db, agent_with_config_and_schedule.agent)
+
+    # Update the agent_id of schedule before scheduling the agent
+    agent_with_config_and_schedule.schedule.agent_id = db_agent.id
+
+    # Then, schedule the agent
+    schedule_id = AgentSchedule.schedule_agent(db.session, agent_with_config_and_schedule.schedule)
```
Comment (TransformerOptimus) [inline superagi/controllers/agent.py:203]: why do we need function for this? directly create here

File: superagi/helper/agent_schedule_helper.py [hunk 8]
```diff
@@ -0,0 +1,113 @@
...
+from datetime import datetime, timedelta
+from superagi.helper.time_helper import parse_interval_to_seconds, check_time_difference
+import pytz
+
+engine = connect_db()
+Session = sessionmaker(bind=engine)
+
+class RunAgentSchedule:
+
+    @staticmethod
```
Comment (TransformerOptimus) [inline superagi/helper/agent_schedule_helper.py:14]: why all of these are static methods. Split it into classes.

File: superagi/helper/agent_schedule_helper.py [hunk 9]
```diff
@@ -0,0 +1,113 @@
...
+                    num_intervals_passed = time_diff.total_seconds() // interval_in_seconds  
+                    updated_next_scheduled_time = agent.start_time + timedelta(seconds=(interval_in_seconds * (num_intervals_passed + 1)))
+                    agent.next_scheduled_time = updated_next_scheduled_time
+                else:
+                    agent.status = "TERMINATED"
+                session.commit()
+        session.close()
+
+    @staticmethod
+    def should_execute_and_remove_agent(agent, interval):
```
Comment (TransformerOptimus) [inline superagi/helper/agent_schedule_helper.py:51]: make this private mthod.

File: superagi/helper/agent_schedule_helper.py [hunk 10]
```diff
@@ -0,0 +1,113 @@
...
+        elif expiry_date is not None and datetime.now() < expiry_date:
+            return True, False
+        elif expiry_runs != -1 and current_runs < expiry_runs:
+            return True, False
+        if (expiry_date is not None and datetime.now() >= expiry_date) or (expiry_runs != -1 and current_runs >= expiry_runs):
+            return False, True
+        return False, False
+
+    @staticmethod
+    def execute_schedule(should_execute_agent, should_remove_agent, interval_in_seconds, session, agent, agent_name):
```
Comment (TransformerOptimus) [inline superagi/helper/agent_schedule_helper.py:68]: make it private method

File: superagi/helper/agent_schedule_helper.py [hunk 11]
```diff
@@ -0,0 +1,113 @@
...
+            agent.current_runs = agent.current_runs + 1
+
+            if agent.recurrence_interval and not should_remove_agent:
+                next_scheduled_time = agent.next_scheduled_time + timedelta(seconds=interval_in_seconds)
+                agent.next_scheduled_time = next_scheduled_time
+
+                session.commit()
+
+    @staticmethod
+    def remove_completed_agents(agents_to_remove, session):
```
Comment (TransformerOptimus) [inline superagi/helper/agent_schedule_helper.py:81]: make it private metjod

File: superagi/jobs/agent_executor.py [hunk 12]
```diff
@@ -292,3 +293,27 @@ def handle_wait_for_permission(self, agent_execution, spawned_agent, session):
         session.add(agent_execution_feed)
         agent_execution.status = "RUNNING"
         session.commit()
+
+
+class ScheduledAgentExecutor:
+
+    @staticmethod
```
Comment (TransformerOptimus) [inline superagi/jobs/agent_executor.py:300]: remove static

File: superagi/jobs/agent_executor.py [hunk 13]
```diff
@@ -292,3 +293,27 @@ def handle_wait_for_permission(self, agent_execution, spawned_agent, session):
         session.add(agent_execution_feed)
         agent_execution.status = "RUNNING"
         session.commit()
+
+
+class ScheduledAgentExecutor:
```
Comment (TransformerOptimus) [inline superagi/jobs/agent_executor.py:298]: why this class here?

File: superagi/models/project.py [hunk 14]
```diff
@@ -55,3 +56,9 @@ def find_or_create_default_project(cls, session, organisation_id):
         else:
             default_project = project
         return default_project
+    
+    @staticmethod
```
Comment (TransformerOptimus) [inline superagi/models/project.py:60]: make it @clasmethod

**Review by ak-gupta89 — APPROVED** [review]

**Review by TransformerOptimus — APPROVED** [review]

**Pull request merged**